test/boost/gcp_object_storage_test.cc

```diff
@@ -0,0 +1,334 @@
+/*
+    Simple test of GCP object storage provider. Uses either real or local, fake, endpoint.
+
+    Note: the above text blobs are service account credentials, including private keys. 
+    _Never_ give any real priviledges to these accounts, as we are obviously exposing them here.
+
+    User1 is assumed to have permissions to read/write the bucket
+    User2 is assumed to only have permissions to read the bucket, but permission to 
+    impersonate User1.
+
+    Note: fake gcp storage does not have any credentials or permissions, so
+    for testing with such, leave them unset to skip those tests.
+
+    This test is parameterized with env vars:
+    * ENABLE_GCP_STORAGE_TEST - set to non-zero (1/true) to run
+    * GCP_STORAGE_ENDPOINT - set to endpoint host. default is https://storage.googleapis.com
+    * GCP_STORAGE_PROJECT - project in which to create bucket (if not specified)
+    * GCP_STORAGE_USER_1_CREDENTIALS - set to credentials file for user1
+    * GCP_STORAGE_USER_2_CREDENTIALS - set to credentials file for user2
+    * GCP_STORAGE_BUCKET - set to test bucket
+*/
+
+struct gcp_test_env {
+    std::string endpoint;
+    std::string project;
+    std::string bucket; 
+    std::string user_1_creds;
+    std::string user_2_creds;
+    // for test
+    std::vector<std::string> objects_to_delete;
+};
+
+static std::string get_var_or_default(const char* var, std::string_view def) {
+    const char* val = std::getenv(var);
+    if (val == nullptr) {
+        return std::string(def);
+    }
+    return val;
+}
+
+using test_func = std::function<future<>(const tmpdir&, storage::client&, gcp_test_env&)>;
+
+static future<std::optional<google_credentials>> credentials(const std::string& source) {
+    try {
+        if (!source.empty()) {
+            co_return co_await google_credentials::from_file(source);
+        }
+        co_return co_await google_credentials::get_default_credentials();
+    } catch (...) {    
+        BOOST_TEST_MESSAGE(fmt::format("Warning: could not load {} credentials: {}", source, std::current_exception()));
+    }
+    co_return std::nullopt; // empty, useless. will work on fake server 
+}
+
+namespace fs = std::filesystem;
+namespace tp = tests::proc;
+
+static future<tp::process_fixture> start_fake_gcs_server(const tmpdir& tmp) {
+    std::vector<std::string> params({
+        "", "-filesystem-root", (tmp.path() / "objects").string()
+    });
+
+    auto exec = tp::find_file_in_path("fake-gcs-server");
+    if (exec.empty()) {
+        std::string image = "fsouza/fake-gcs-server";
+        exec = tp::find_file_in_path("docker");
+        if (exec.empty()) {
+            exec = tp::find_file_in_path("podman");
+            image = "docker.io/fsouza/fake-gcs-server";
+        }
+        if (exec.empty()) {
+            throw std::runtime_error("Could not find fake-gcs-server or docker.");
+        }
+        params = {
+            "",
+            "run", "--rm", "--name", "fake-gcs-server", "-p", "4443:4443",
```

Comment:
Need to allocate either a random bind address or a random port so parallel runs can work.

Reply:
The server is immaterial here since you bind whatever port in the container to a port in the host (which is also a container under dbuild).